File: common/src/main/java/juuxel/adorn/util/Colors.java
package juuxel.adorn.util;

public final class Colors {
    public static final int BLACK = 0xFF_000000;
    public static final int WHITE = 0xFF_FFFFFF;
    public static final int SCREEN_TEXT = 0xFF_404040;
    public static final int TRANSPARENT = 0x00_000000;

    public static int color(int rgb) {
        return color(rgb, 0xFF);
    }

    public static int color(int rgb, int alpha) {
        return alpha << 24 | rgb;
    }

    public static int color(int rgb, float alpha) {
        return color(rgb, (int) (alpha * 255f));
    }

    public static int color(float red, float green, float blue) {
        return color(red, green, blue, 1f);
    }

    public static int color(float red, float green, float blue, float alpha) {
        return color((int) (red * 255f), (int) (green * 255f), (int) (blue * 255f), (int) (alpha * 255f));
    }

    public static int color(int red, int green, int blue, int alpha) {
        return (alpha & 0xFF) << 24 | (red & 0xFF) << 16 | (green & 0xFF) << 8 | blue & 0xFF;
    }

    public static float redOf(int argb) {
        return ((argb >> 16) & 0xFF) / 255f;
    }

    public static float greenOf(int argb) {
        return ((argb >> 8) & 0xFF) / 255f;
    }

    public static float blueOf(int argb) {
        return (argb & 0xFF) / 255f;
    }

    public static float alphaOf(int argb) {
        return ((argb >> 24) & 0xFF) / 255f;
    }
}

File: common/src/main/kotlin/juuxel/adorn/util/Colors.kt
package juuxel.adorn.util

@JvmOverloads
fun color(rgb: Int, alpha: Int = 0xFF) =
    alpha shl 24 or rgb

fun color(rgb: Int, alpha: Float) = color(rgb, (alpha * 255f).toInt())

fun color(red: Float, green: Float, blue: Float, alpha: Float = 1f): Int =
    color((red * 255f).toInt(), (green * 255f).toInt(), (blue * 255f).toInt(), (alpha * 255f).toInt())

fun color(red: Int, green: Int, blue: Int, alpha: Int = 0xFF): Int =
    ((alpha and 0xFF) shl 24) or ((red and 0xFF) shl 16) or ((green and 0xFF) shl 8) or (blue and 0xFF)

object Colors {
    @JvmField
    val BLACK = color(0x000000)

    @JvmField
    val WHITE = color(0xFFFFFF)

    @JvmField
    val SCREEN_TEXT = color(0x404040)

    @JvmField
    val TRANSPARENT = color(0x000000, alpha = 0)

    @JvmStatic
    fun redOf(argb: Int): Float = ((argb shr 16) and 0xFF) / 255f

    @JvmStatic
    fun greenOf(argb: Int): Float = ((argb shr 8) and 0xFF) / 255f

    @JvmStatic
    fun blueOf(argb: Int): Float = (argb and 0xFF) / 255f

    @JvmStatic
    fun alphaOf(argb: Int): Float = ((argb shr 24) and 0xFF) / 255f
}
